lib/jekyll/cache.rb

```diff
@@ -0,0 +1,175 @@
+# frozen_string_literal: true
+
+require "digest"
+require "fileutils"
+require "pstore"
+
+module Jekyll
+  class Cache
+    extend Forwardable
+
+    # rubocop:disable Style/ClassVars
+    @@caches = {}
+    @@disk_cache_enabled = true
+
+    # Get an existing named cache, or create a new one if none exists
+    #
+    # name - name of the cache
+    #
+    # Returns nothing.
+    def initialize(name)
+      @@base_dir ||= File.expand_path(".jekyll-cache/Jekyll/Cache")
+      @cache = @@caches[name] ||= {}
+      @name = name.gsub(%r![^\w\s-]!, "-")
+      FileUtils.mkdir_p(path_to) if @@disk_cache_enabled
+    end
+
+    # Disable Marshaling cached items to disk
+    def self.disable_disk_cache!
+      @@disk_cache_enabled = false
+    end
+    # rubocop:enable Style/ClassVars
+
+    # Clear all caches
+    def self.clear
+      delete_cache_files
+      @@caches.each_value(&:clear)
+    end
+
+    # Clear this particular cache
+    def clear
+      delete_cache_files
+      @cache.clear
+    end
+
+    # Retrieve a cached item
+    # Raises if key does not exist in cache
+    #
+    # Returns cached value
+    def [](key)
+      return @cache[key] if @cache.key?(key)
+      path = path_to(hash(key))
+      if @@disk_cache_enabled && File.file?(path) && File.readable?(path)
+        @cache[key] = load(path)
+      else
+        raise
+      end
+    end
+
+    # Add an item to cache
+    #
+    # Returns nothing.
+    def []=(key, value)
+      @cache[key] = value
+      if @@disk_cache_enabled
+        path = path_to(hash(key))
+        dump(path, value)
+      end
+    end
+
+    # If an item already exists in the cache, retrieve it
+    # Else execute code block, and add the result to the cache, and return that
+    #   result
+    def getset(key)
+      self[key]
+    rescue StandardError
+      value = yield
+      self[key] = value
+      value
+    end
+
+    # Remove one particular item from the cache
+    #
+    # Returns nothing.
+    def delete(key)
+      @cache.delete(key)
+      if @@disk_cache_enabled
+        path = path_to(hash(key))
+        File.delete(path)
+      end
+    end
+
+    # Check if `key` already exists in this cache
+    #
+    # Returns true if key exists in the cache, false otherwise
+    def key?(key)
+      # First, check if item is already cached in memory
+      return true if @cache.key?(key)
+      # Otherwise, it might be cached on disk
+      # but we should not consider the disk cache if it is disabled
+      return false unless @@disk_cache_enabled
+      path = path_to(hash(key))
+      File.file?(path) && File.readable?(path)
+    end
+
+    # Compare the current config to the cached config
+    # If they are different, clear all caches
+    #
+    # Returns nothing.
+    def self.clear_if_config_changed(config)
+      config = config.inspect
+      cache = Jekyll::Cache.new "Jekyll::Cache"
```

Comment:
Would delegating `Jekyll::Cache.new "Jekyll::Cache"` to a dedicated method called `:initialize_self` be over-engineering?
```ruby
def initialize_self
  Jekyll::Cache.new("Jekyll::Cache") # or simply `new("Jekyll::Cache")`
end
```

Reply:
🤷‍♂️ I'm not sure. This is the only place we are using it.

To be consistent with the way I am using the cache in other places, we could call the method `cache`, but it might be odd to have a `cache` method in the `Cache` class.

![Yo Dawg, I hear you like caches...](https://user-images.githubusercontent.com/251545/44226882-deaded80-a156-11e8-9115-7515f3fd989e.jpg)

Reply:
I think this usage is fine. There is such a thing as over-engineering.